### ISSUE TITLE
Composite FK annotation

### ISSUE THREAD

**Issue by eldhenn** [dotnet/efcore#31881]
We have PrimaryKeyAttribute, thank you.  But AFAICS there is no way to map foreing key same way.

**Comment by ajcvickers** [comment]
@eldhenn `ForeignKeyAttribute` can be used for that. For example:

```C#
[PrimaryKey(nameof(Id1), nameof(Id2))]
public class Blog
{
    public Guid Id1 { get; set; }
    public int Id2 { get; set; }
    public string? Name { get; set; }
    public List<Post> Posts { get; } = new();
}

public class Post
{
    public Guid Id { get; set; }
    public string? Title { get; set; }
    public Guid? BlogIdA { get; set; }
    public int? BlogIdB { get; set; }
   
    [ForeignKey("BlogIdA,BlogIdB")]
    public Blog? Blog { get; set; }
}
```

**Comment by eldhenn** [comment]
> "BlogIdA,BlogIdB"

Eww. No any checkings at all...

**Comment by piskov** [comment]
@eldhenn you can use as a bullet-proof `[ForeignKey($"{nameof(a)}, {nameof(b)}, {nameof(c)}")]

@ajcvickers may I ask you to reconsider? You could always have `[PrimaryKey($"{nameof(a)}, {nameof(b)}")]` yet you chose to add params, no?

So it’s just a matter of consistency. These two should be the same:

```
PrimaryKeyAttribute(String, String[])
ForeignKeyAttribute(String, String[])
```

… instead of current
```
PrimaryKeyAttribute(String, String[])
ForeignKeyAttribute(String)
```

**Comment by roji** [comment]
@piskov the problem here is mainly that [`PrimaryKey`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.primarykeyattribute?view=efcore-7.0) is part of EF Core itself; we own it, it only impacts EF (no other ORM/package) and so it's easy to make changes.

In contrast, [`ForeignKey`](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.schema.foreignkeyattribute?view=net-7.0) is part of .NET itself (System.ComponentModel.DataAnnotations), so it's not EF-specific, and evolving that can be quite tricky. We don't believe that such "generic" data schema attributes are a good idea in general, but it is what it is.

**Comment by piskov** [comment]
@roji got it. Any chance you can point me in the direction where one could propose changes to data annotations? I suppose getting your own `ForeignKey` for example via inheritance is not an option for you, as usually both namespaces will be imported, hence the clash.

But consistency, predictability, and discovery are paramount. In case of data annotation no-one could “deduce” that `[ForeignKey("Id1,Id2")]` would work (“denotes a _property_” in the docs)

**Comment by roji** [comment]
@piskov technically the place for that would be http://github.com/dotnet/runtime; we'd still probably be involved in making the decision about that.

I agree that the situation with the comma isn't ideal, and had we designed that attribute today, we obviously would not have done things this way. However, things are the way they are, and making changes in the existing attribute can be tricky.

For example, we'd presumably introduce a new overload for [ForeignKey] which accepts two or more string parameters (with params). While EF Core could be adjusted to recognize this new feature, other ORMs recognizing [ForeignKey] would not (e.g. the older EF6, nHibernate), leading to fragmented, inconsistent story. This, by the way, is why we think that having such "general" schema attributes which aren't tied to any specific ORM/package is a mistake; the [PrimaryKey] attribute is part of EF Core, is only recognized by it, and therefore has clear semantics and can be safely evolved.

In addition, we'd be in a situation where the single-string overload supports commas inside as a separator, but the newer one does not, which is inconsistent.

The bottom line here is that even if the current situation isn't ideal, it's workable, and there's probably not enough actual value to try to improve [ForeignKey].
